# Preprocessing of scraped recipes

### Import libraries

In [ ]:
# pip install tensorflow==2.1

In [2]:
import pandas as pd
import re
import pickle
import numpy as np
pd.set_option('display.max_colwidth', None)

from collections import Counter
import string
from itertools import chain

# import tensorflow as tf
# from keras.layers import Input
# from keras.models import Model
# from keras.models import Sequential
# from keras.layers import Activation, Dense, LSTM, Embedding, TimeDistributed
# tf.keras.backend.clear_session()  # For easy reset of notebook s

### Load the data

In [16]:
# json_file = 'C:/Users/galin/PycharmProjects/aiAssignment/data/recipes.pkl'
# set_json = pickle.load(open(json_file, 'rb'))
# type(set_json)

list

In [17]:
# filename='C:/Users/galin/PycharmProjects/aiAssignment/data/recipes.pkl'
# df = pd.DataFrame(pd.read_pickle(filename))
# df.head()

Recipe  \
0                                                                                                                                                                                                                [2 cups diced strawberries, ½ cup frozen raspberries, thawed, with their juices, Pinches of cane sugar]   
1  [½ cup warm water, 110°, 1 (¼-ounce) package active dry yeast, 1 teaspoon cane sugar, ⅓ cup melted coconut oil, plus more for brushing, ½ cup almond milk, at room temperature, ⅓ cup cane sugar, 1 teaspoon sea salt, 2¾ cups all-purpose flour, plus more for kneading, ½ cup brown sugar, 1½ tablespoons cinnamon]   
2                                                                                                                                                 [2 cups cubed butternut squash, 2 cups halved Brussels sprouts, 2 cups cauliflower florets, 2 carrots, cut into 1-inch pieces, 1 to 2 turnips, cut into 1-inch pieces]   
3                                                                                                                       [1¾ cups warm water, 105° to 115°F, 1 (¼-ounce) package active dry yeast, (2¼ teaspoons), 1 tablespoon cane sugar, 3½ cups all-purpose flour, plus more for kneading, 1½ cups whole wheat flour]   
4                                                                                                                                                                                                                    [2 cups all-purpose flour, 1½ teaspoons baking powder, ½ teaspoon baking soda, 1 teaspoon sea salt]   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            INSTRUCTIONS  
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Mak

### Check the shape and info

In [19]:
# len(set_json)

1035

In [20]:
# set_json

In [21]:
# df.shape

(1035, 2)

<b><br> We have 1035 rows and 2 columns. 
<br>`Recipe` and `INSTRUCTIONS` are our 2 columns.
<br> Also our present columns will be our labels for each sentence. 
<br> Later we will split the data for sentences: 
<br>each row will be a sentence and we will have a label 1- recipe and 0 - instructions.</b>

In [22]:
# df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1035 entries, 0 to 1034
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Recipe        1035 non-null   object
 1   INSTRUCTIONS  1035 non-null   object
dtypes: object(2)
memory usage: 16.3+ KB


<b>Ok, good. <br>Now let's transform our first column `Recipe` :</b>

In [23]:
# recipe_col = df["Recipe"].to_numpy()
# recipe = np.concatenate(recipe_col).reshape(-1,1)
# recipe = pd.DataFrame(recipe, columns=['sentence'])
# recipe.head()

sentence
0                            2 cups diced strawberries
1  ½ cup frozen raspberries, thawed, with their juices
2                                Pinches of cane sugar
3                               ½ cup warm water, 110°
4                 1 (¼-ounce) package active dry yeast

<b>We have to add a column `label`:</b>

In [24]:
# recipe['label'] = np.ones(len(recipe),int)
# recipe.head()

sentence  label
0                            2 cups diced strawberries      1
1  ½ cup frozen raspberries, thawed, with their juices      1
2                                Pinches of cane sugar      1
3                               ½ cup warm water, 110°      1
4                 1 (¼-ounce) package active dry yeast      1

In [25]:
# recipe.to_numpy().shape

(4160, 2)

<b>Great! It's time to prepare the `INSTRUCTIONS` column</b>

In [26]:
# instr_col = df["INSTRUCTIONS"].str.split('\n').to_numpy()
# instr = np.concatenate(instr_col).reshape(-1,1)
# instr = pd.DataFrame(instr, columns=['sentence'])
# instr.head()

sentence
0                                                                               Make the macerated berries: In a medium bowl, combine the strawberries, raspberries, and a few pinches of sugar. Set aside for 10 minutes for the berries to soften. Stir before serving.
1                                                                                                                                                                                                                                                                        
2                                                                  Make the French toast: In a large bowl, whisk together the eggs, milk, cinnamon, cardamom, and salt. Dip each slice of bread into the mixture and set the soaked bread aside on a large tray or plate.
3                                                                                                                                                                                                                                                                        
4  Heat a non-stick skillet to medium heat and brush with coconut oil. Add the bread slices and cook until golden brown, about 2 minutes per side. Reduce the heat to low as needed to cook thoroughly without burning. Serve with maple syrup and the macerated berries.

<b>Adding column `label` with value 0</b>

In [27]:
# instr['label'] = np.zeros(len(instr),int)
# instr

sentence  \
0                                                                                                                        Make the macerated berries: In a medium bowl, combine the strawberries, raspberries, and a few pinches of sugar. Set aside for 10 minutes for the berries to soften. Stir before serving.   
1                                                                                                                                                                                                                                                                                                                    
2                                                                                                           Make the French toast: In a large bowl, whisk together the eggs, milk, cinnamon, cardamom, and salt. Dip each slice of bread into the mixture and set the soaked bread aside on a large tray or plate.   
3                                                                                                                                                                                                                                                                                                                    
4                                           Heat a non-stick skillet to medium heat and brush with coconut oil. Add the bread slices and cook until golden brown, about 2 minutes per side. Reduce the heat to low as needed to cook thoroughly without burning. Serve with maple syrup and the macerated berries.   
...                                                                                                                                                                                                                                                                                                            ...   
7300                                                                                                                                                 For "fettucine" noodles using a mandoline and a knife: Slice thin zucchini planks with the mandoline, and then cut those planks into fettuccine-sized strips.   
7301                                                                                                                                                                                                                                                                                                                 
7302                                                                                                                                                                                            For "pappardelle" noodles using a regular vegetable peeler: Use the peeler to simply peel thin strips of zucchini.   
7303                                                                                                                                                                                                                                                                                                                 
7304  Serve the zucchini noodles raw with a warm sauce (the heat of the sauce will gently cook the noodles without making them mushy). Or heat a skillet to medium heat, brush with olive oil, add the noodles and heat 1 minute, or until just warmed through. Remove and serve with desired sauces and toppings.   

      label  
0         0  
1         0  
2         0  
3         0  
4         0  
...     ...  
7300      0  
7301      0  
7302      0  
7303      0  
7304      0  

[7305 rows x 2 columns]

In [28]:
# instr.to_numpy().shape

(7305, 2)

In [29]:
# data = pd.DataFrame(np.concatenate((instr.to_numpy(), recipe.to_numpy())), columns=['paragraph','label'])
# data.head()

paragraph  \
0                                                                               Make the macerated berries: In a medium bowl, combine the strawberries, raspberries, and a few pinches of sugar. Set aside for 10 minutes for the berries to soften. Stir before serving.   
1                                                                                                                                                                                                                                                                           
2                                                                  Make the French toast: In a large bowl, whisk together the eggs, milk, cinnamon, cardamom, and salt. Dip each slice of bread into the mixture and set the soaked bread aside on a large tray or plate.   
3                                                                                                                                                                                                                                                                           
4  Heat a non-stick skillet to medium heat and brush with coconut oil. Add the bread slices and cook until golden brown, about 2 minutes per side. Reduce the heat to low as needed to cook thoroughly without burning. Serve with maple syrup and the macerated berries.   

  label  
0     0  
1     0  
2     0  
3     0  
4     0

In [30]:
# data.shape

(11465, 2)

In [31]:
# data.label[data.label == 1].count()

4160

In [32]:
# data.label[data.label == 0].count()

7305

In [33]:
# data.isna().sum()

paragraph    0
label        0
dtype: int64

##### no null - good, check the proportion

In [34]:
# round(data['label'].value_counts()/len(data)*100,2)

0    63.72
1    36.28
Name: label, dtype: float64

##### drop duplicated and empty rows

In [35]:
# data = data[data.paragraph != '']


In [36]:
# data = data.drop_duplicates().reset_index(drop=True)
# data.tail()

paragraph label
6857                      1 cup diced tomato     1
6858                 ½ cup diced white onion     1
6859                  ¼ cup chopped cilantro     1
6860  1 large sweet potato (just under 1 lb)     1
6861                   ½ medium onion, diced     1

##### check the proportion again

In [37]:
# round(data['label'].value_counts()/len(data)*100,2)

0    56.25
1    43.75
Name: label, dtype: float64

##### Chenged not much, but +6.2%  of sentences in  `ingredients` and -6.2% in `recipe` after droping duplacates
##### We have a little bit imballanced data set
##### Replace `.` to token `zzdot` and split on it

In [39]:
# DIGIT_RX = "(\d+([\.|,]\d+)?[\w]*[\s|-])|[^A-Za-z\,()\.'\-: ]{1,7}"
# SIMBOL_RX ="[/(/)\-/*/,]|[^ -~]"
# DOT_RX = "\.|\:"

# def replace_numbers_str(series):
#         """
#         replace numbers expression (11 ; 11,00;  1111.99; 23-th 25-,1/2,¼) with tag ' zNUM ',
#         . and : replace with ' zDOT '
#         :param series: pandas series of text strings
#         :return: series with replaced numbers
#         """
#         new_series1 = re.sub(DIGIT_RX," zNUM ",series)
#         new_series2 = re.sub(SIMBOL_RX,' ',new_series1)
#         new_series = re.sub(DOT_RX," zDOT",new_series2)
#         return new_series

# data['paragraph'] = data.paragraph.apply(replace_numbers_str)
# data.tail(15)

paragraph label
6847                                   zNUM  cup dried cranberries     1
6848   zNUM sticks  zNUM   zNUM tablespoons butter or vegan butter     1
6849                             zNUM ounces bittersweet chocolate     1
6850                                                zNUM egg yolks     1
6851                                      zNUM cups powdered sugar     1
6852                           zNUM cup all purpose or spelt flour     1
6853                         zNUM to  zNUM tablespoons coconut oil     1
6854                                        zNUM  cup popping corn     1
6855     zNUM tablespoons earth balance butter  or regular butter      1
6856                              Corn tortillas or tortilla chips     1
6857                                         zNUM cup diced tomato     1
6858                                   zNUM  cup diced white onion     1
6859                                    zNUM  cup chopped cilantro     1
6860                 zNUM large sweet potato  just under  zNUM lb      1
6861                                     zNUM  medium onion  diced     1

In [3]:
# load data
filename='./recipes.pkl'
df = pd.DataFrame(pd.read_pickle(filename))

#transform recioe to array and give a label 1
recipe_col = df["Recipe"].to_numpy()
recipe = np.concatenate(recipe_col).reshape(-1,1)
recipe = np.hstack((recipe,np.ones(len(recipe),int).reshape(-1,1)))

#transform instructions to array and give a label 0
instr_col = df["INSTRUCTIONS"].str.split('\n').to_numpy()
instr = np.concatenate(instr_col).reshape(-1,1)
instr = np.hstack((instr,np.zeros(len(instr),int).reshape(-1,1)))

#forming a full data array with labels
data = np.concatenate((instr, recipe))
print('shape all of paragraphs data matrix',data.shape)

# remove duplicates
unique = np.unique(data, axis=0)
print('shape without duplicates',unique.shape)

#remove empty string rows(from table)
unique = np.delete(unique,np.where(unique == ''),axis=0)
print('shape without empty string rows',unique.shape)

#constants regex
DIGIT_RX = "(\d+([\.|,]\d+)?[\w]*[\s|-])|[^A-Za-z\,()\.'\-: ]{1,7}"
SIMBOL_RX ="[/(/)\-/*/,]|[^ -~]"
DOT_RX = "\.{1,4}|\:" 


def replace_numbers_str(series):
        """
        replace numbers expression (11 ; 11,00;  1111.99; 23-th 25-,1/2,¼) with tag ' zNUM ',
        . and : replace with ' zDOT '
        :param series: pandas series of text strings
        :return: series with replaced numbers
        """
        new_series1 = re.sub(DIGIT_RX," zNUM ",series)
        new_series2 = re.sub(SIMBOL_RX,' ',new_series1)
        new_series = re.sub(DOT_RX," zDOT ",new_series2)
        return new_series
    
#transform to pandas -> easy to clean    
data = pd.DataFrame(unique, columns=['paragraph', 'label'])
data['paragraph'] = data.paragraph.apply(replace_numbers_str)
data[1:4]

shape all of paragraphs data matrix (11465, 2)
shape without duplicates (6863, 2)
shape without empty string rows (6862, 2)


paragraph  \
1   Optional zDOT  if you're making okra  drizzle with olive oil  salt  zNUM  pepper and roast at the same time   for about  zNUM minutes or until they start to brown  zDOT    
2                                                                                                                                   also good on pasta  in salads  etc zDOT     
3                                                                                                         optional   I served mine on top of a few handfuls of arugula  zDOT    

  label  
1     0  
2     0  
3     0

In [4]:
def count_paragraph_stats(series):
    """Count number of sentences in the paragraph
        :param series: pandas series of text strings
        :return: int count of sentences in the string (if no dots -> count = 1)"""
    sent_count = series.count(' zDOT ')
    if sent_count == 0:
        sent_count = 1
    return sent_count

def num_count(series):
    """Count number of sentences in the paragraph
        :param series: pandas series of text strings
        :return: int count of sentences in the string"""
    return series.count('zNUM') 

data['sent_count'] = data['paragraph'].apply(count_paragraph_stats)
data['num_count'] = data['paragraph'].apply(num_count)
data.head(50)

paragraph  \
0                                                                                                                                                                                                                                            Alternatively  omit the sparkling water  shake everything in a cocktail shaker with ice and serve in a rocks glass  zDOT    
1                                                                                                                                                                                            Optional zDOT  if you're making okra  drizzle with olive oil  salt  zNUM  pepper and roast at the same time   for about  zNUM minutes or until they start to brown  zDOT    
2                                                                                                                                                                                                                                                                                                                            also good on pasta  in salads  etc zDOT     
3                                                                                                                                                                                                                                                                                                  optional   I served mine on top of a few handfuls of arugula  zDOT    
4                                                                                                                                                                                                                                                                                         optional step  Mix your yogurt sauce together and pop it in the fridge zDOT    
5                                                                                                                                optional  Pan sear your tofu slabs zDOT  Press the water out with paper towels  drizzle with a bit of soy sauce zDOT  Heat oil in a skillet  add the tofu  and cook for a few minutes per side zDOT  Remove and chop into cubes zDOT    
6                                                                                                                                                                                                                                    optional  Toss carrot ribbons with a splash of rice vinegar and a few pinches of salt zDOT  Let marinate until ready to use zDOT    
7                                                                                                                                                                                                                                                                                                    or  zNUM cups uncooked shells and sauce from here  here  or here    
8                                                                                                                                                                                                                                                                              or about  zNUM  zNUM fresh roma tomatoes  roast them with the onions  zNUM  peel skins    
9                                                                                                                                                                                                                                                                                                                               or good cold pressed watermelon juice    
10                                                                                                                                                                                                                                                                                                                                           reserve some pasta water    
11   

##### Count words with stopwords and symbols

In [ ]:
!pip install -U spacy
# !pip install -U spacy[cuda92]

# import spacy
# ! python -m spacy download en

ERROR: Exception:
Traceback (most recent call last):
  File "C:\Users\galin\Anaconda3\lib\site-packages\pip\_vendor\urllib3\response.py", line 438, in _error_catcher
    yield
  File "C:\Users\galin\Anaconda3\lib\site-packages\pip\_vendor\urllib3\response.py", line 519, in read
    data = self._fp.read(amt) if not fp_closed else b""
  File "C:\Users\galin\Anaconda3\lib\site-packages\pip\_vendor\cachecontrol\filewrapper.py", line 62, in read
    data = self.__fp.read(amt)
  File "C:\Users\galin\Anaconda3\lib\http\client.py", line 458, in read
    n = self.readinto(b)
  File "C:\Users\galin\Anaconda3\lib\http\client.py", line 502, in readinto
    n = self.fp.readinto(b)
  File "C:\Users\galin\Anaconda3\lib\socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "C:\Users\galin\Anaconda3\lib\ssl.py", line 1241, in recv_into
    return self.read(nbytes, buffer)
  File "C:\Users\galin\Anaconda3\lib\ssl.py", line 1099, in read
    return self._sslobj.read(len, buffer)
soc

In [3]:
from spacy.lang.en import English
from spacy.language import Language
from spacy.tokenizer import Tokenizer
from spacy.tokens import Doc, Token
import utils
from utils.pipeline import Pipeline
from nlp.stop_words import STOP_WORDS, QUESTION_WORDS

ModuleNotFoundError: No module named 'spacy'